caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalCacheFactory.java

```diff
@@ -80,10 +83,22 @@ static String getClassName(Caffeine<?, ?> builder) {
 
   static <K, V> BoundedLocalCache<K, V> loadFactory(Caffeine<K, V> builder,
       @Nullable AsyncCacheLoader<? super K, V> cacheLoader, boolean async, String className) {
+    var constructor = CONSTRUCTORS.get(className);
+    if (constructor == null) {
+      constructor = CONSTRUCTORS.computeIfAbsent(className, LocalCacheFactory::newConstructor);
+    }
+    try {
+      return (BoundedLocalCache<K, V>) constructor.invokeExact(builder, cacheLoader, async);
+    } catch (Throwable t) {
+      throw new IllegalStateException(className, t);
+    }
+  }
+
+  static MethodHandle newConstructor(String className) {
     try {
-      Class<?> clazz = Class.forName(className);
-      MethodHandle handle = LOOKUP.findConstructor(clazz, FACTORY);
-      return (BoundedLocalCache<K, V>) handle.invoke(builder, cacheLoader, async);
+      var clazz = LOOKUP.findClass(LocalCacheFactory.class.getPackageName() + "." + className);
+      var constructor = LOOKUP.findConstructor(clazz, FACTORY);
+      return constructor.asType(constructor.type().changeReturnType(BoundedLocalCache.class));
```

Comment:
Perfect. Thanks. Sorry for proposing slightly wrong code. In fact, you can only change return type on MethodType not MethodHandle. 😉 Your code looks perfect to me.